### PR TITLE
Feat: support VirtualXposed.

### PR DESCRIPTION
**Pull request by tiann**
Hi M66B,

I am the author of [VirtualXposed](https://github.com/android-hacker/VirtualXposed), I have added support for XPrivacyLua recently in VirtualXposed, but i have to modify the source code of XPrivacyLua. This PR makes XPrivacyLua work in VirtualXposed.

We have discussed it in https://github.com/android-hacker/VirtualXposed/issues/7, I am sorry for my recklessness :(

I have tested it and i think it works well, here are some screenshots:

restrict the camera permission in XPrivacyLua
image

### CODE REVIEW

**Comment by M66B** [comment]
Apart from that the solution is insecure because the privacy database will be accessible to all apps,  it will not work for system apps and components, so this cannot provide real privacy. Therefore I don't want to support this.

**Comment by tiann** [comment]
Yes, it can not work for system apps and components, but it can provide privacy for apps installed in VirtualXposed. Some device is difficult or impossible to install Xposed,  this solution can provide restricted privacy for those users.

As for the permission of database, i find a way to keep it stay in /data/system/xlua/xlua.db, you can refer it : https://github.com/M66B/XPrivacyLua/pull/4

The modification doesn't affect the original function of Xposed users, and it may bring a lot of users for XPrivacyLua. I think it is a win-win for XPrivacyLua and VirtualXposed, could you please think about it again?

**Comment by M66B** [comment]
You cannot store the privacy database in a safe place without root permissions and it will still never be a real privacy solution, so my opinion stays unchanged.

**Comment by tiann** [comment]
but If the device have root, the database is unsafe, too.

**Comment by M66B** [comment]
No, it isn't, only if the user grants root permissions to an app.

**Comment by tiann** [comment]
OK, I will maintain a fork.

**Comment by M66B** [comment]
That is okay with me, as long as you support it too. This is another reason I don't want to support XPrivcayLua on VirtualXposed: I don't want to start answering lots of questions about why things don't work, like system apps, same as in the VirtualXposed main thread.

File: app/src/main/java/eu/faircode/xlua/XContentProvider.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        return false;
+    }
+
+    @Nullable
+    @Override
+    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
+        try {
+            return XProvider.query(getContext(), projection[0].split("\\.")[1], selectionArgs);
+        } catch (RemoteException e) {
+            return null;
```
Comment (M66B) [inline app/src/main/java/eu/faircode/xlua/XContentProvider.java:40]: No error logging

File: app/src/main/java/eu/faircode/xlua/XProvider.java [hunk 2]
```diff
@@ -104,6 +104,7 @@ static Bundle call(Context context, String method, Bundle extras) throws RemoteE
                     result = getGroups(context, extras);
                     break;
                 case "assignHooks":
+                    Log.i(TAG, "assignHooks...");
```
Comment (M66B) [inline app/src/main/java/eu/faircode/xlua/XProvider.java:107]: Unneeded logging

File: app/src/main/java/eu/faircode/xlua/XProvider.java [hunk 3]
```diff
@@ -1146,6 +1149,15 @@ private static SQLiteDatabase getDatabase() throws Throwable {
                         "system" + File.separator +
                         "xlua" + File.separator +
                         "xlua.db");
+
+        if (Util.isVirtualXposed()) {
+            // In VirtualXposed, we can not access /data/system, use /sdcard/ instead
+            dbFile = new File(
```
Comment (M66B) [inline app/src/main/java/eu/faircode/xlua/XProvider.java:1155]: Insecure storage of privacy database

File: app/src/main/java/eu/faircode/xlua/ActivityMain.java [hunk 4]
```diff
@@ -82,7 +81,7 @@ protected void onCreate(Bundle savedInstanceState) {
             Snackbar snackbar = Snackbar.make(findViewById(android.R.id.content), getString(R.string.msg_no_service), Snackbar.LENGTH_INDEFINITE);
 
             final Intent intent = getPackageManager().getLaunchIntentForPackage("de.robv.android.xposed.installer");
-            if (intent.resolveActivity(getPackageManager()) != null)
+            if (intent == null || intent.resolveActivity(getPackageManager()) != null)
```
Comment (M66B) [inline app/src/main/java/eu/faircode/xlua/ActivityMain.java]: There are more references to Xposed, like logging: did you check if that works okay?

File: app/src/main/java/eu/faircode/xlua/ActivityMain.java [hunk 5]
```diff
@@ -24,9 +24,9 @@
 import android.content.Intent;
 import android.content.SharedPreferences;
 import android.content.pm.PackageManager;
-import android.content.pm.ResolveInfo;
 import android.content.res.Configuration;
 import android.net.Uri;
+import android.os.Bundle;
```
Comment (M66B) [inline app/src/main/java/eu/faircode/xlua/ActivityMain.java]: Not needed import.

**Comment by M66B** [comment]
If you can find a way to store the privacy database not on public accessible storage, I will consider to merge this. Suggestion: did you try to store the database in the private app data folder? (/data/data/eu.faircode.xlua)

**Comment by Namnodorel** [comment]
Yeah, the private app folder also would've been the first thing to come to my mind if you want to store something relatively safe from other apps without root. Android should keep any other apps from accessing that anyways, so that looks like a good place to put the DB in in "VirtualXposed Mode".

On the note of not restricting system apps: I wouldn't say that is a good reason to not want to have XPL on VirtualXposed. Even with Root, XPL can't cover apps 100% due to native code, and does that make it less advanagous for privacy or even useless? No! Being able to restrict user apps with XPL in VX is still more privacy than ever was possible without an unlocked bootloader, even though it is not _as good_ as the original version.

**Comment by tiann** [comment]
Sorry for reply too late :(

I have found a way to store the database safely:

1. Keep the database file stay in /data/system/xlua, in a non-root environment, this folder can never be accessed by any app.
2. Make an IO redirect for apps in VirtualXposed:
    - When in XPrivacyLua process, redirect /data/system/xlua to VirtualXposed's private app data folder(/data/data/io.va.exposed), this directory can not be accessed from other apps in system.
    - Do nothing with other app in VirtualXposed, if these app access /data/system/xlua, they will access the real /data/system folder, which will never be successful.  
    - When do Xposed hooks in apps, the app just make an IPC to XPrivacyLua's process (XContentProvider)

Here are my commits: https://github.com/android-hacker/VirtualXposed/commit/930961b88ba11ea882b47034a1b21fddbbe6ee77

And i make a new PR which fix the problem M66B reviewed above: https://github.com/M66B/XPrivacyLua/pull/4

As for the questions of users, you can add some FAQs or simply tell those users to contact with me :)

**Comment by M66B** [comment]
The pull request doesn't include routing to the new content provider or am I missing something? In any case it doesn't work for me because

```07-01 13:50:56.044   836  6836 W SettingsProvider: call() with invalid method: xlua```

You have changed VirtualXposed to route the database path to another folder. Why not simply use the app data folder? As it is now a new version of VirtualXposed is required, which will inevitably result in lots of questions.

```
        if (Util.isVirtualXposed())
            dbFile = new File(
                    Environment.getDataDirectory() + File.separator +
                            "data" + File.separator +
                            BuildConfig.APPLICATION_ID + File.separator +
                            "xlua.db");
        else
            dbFile = new File(
                    Environment.getDataDirectory() + File.separator +
                            "system" + File.separator +
                            "xlua" + File.separator +
                            "xlua.db");
```

**Comment by M66B** [comment]
What also won't work is new app installed notifications and restricting by default.

**Comment by tiann** [comment]
I am sorry, I haven't made new release now, the version used by user doesn't support XPL now. To support XPL, i made some other changes, so they must upgrade VX to use it.
[here is test version: VirtualXposed_0.12.5.apk.zip](https://github.com/M66B/XPrivacyLua/files/2152780/VirtualXposed_0.12.5.apk.zip)

> Why not simply use the app data folder?

Yes, this can also work. But as i mentioned above, there are other changes must be made to support XPL, the new release is inevitable, my solution can make XPL stay the same.

1. new app installed notifications: this is not supported in VX now :( but i will support it soon.
2. restricting by default: i don't know this feature before, but i think supporting this do not need to modify XPL anymore.

**Comment by tiann** [comment]
this commit: https://github.com/android-hacker/VirtualXposed/commit/f27f97f2f465dfd4c4b0fb33f6d2f01bb5e2f8ee

redirect the access to Settings's ContentProvider to XPL's XContentProvider.

**Comment by M66B** [comment]
With https://github.com/M66B/XPrivacyLua/commit/2f45dba9b2e7c530af845776bdb4b4ec8f05c821 there is no need for changes in VirtualXposed.

**Comment by M66B** [comment]
In fact, better remove all changes for XPrivacyLua from VirtualXposed to prevent trouble and confusion.

**Comment by tiann** [comment]
Your solution is more sensible and it works fine. I have made a lot of adaptation in VXP for many Xposed modules because authors of the modules don't want to modify their code.

Thank you, M66B :)

**Pull request closed without merging**